### PR TITLE
import: on root folder change set it as last directory by default

### PR DESCRIPTION
**Pull request by phweyland**
Let see if that make more people happy

EDIT. recursive directory checked may be boring i that case ...

### CODE REVIEW

**Comment by elstoc** [comment]
If you're going to do this by default you need to add the root folder to the tree and auto-expand it as I suggested in https://github.com/darktable-org/darktable/pull/8734#issuecomment-825429990.

Also something seems to have broken. I set the root as follows:

Screenshot_2021-04-23_12-45-33

If I try to change the root to `/home/chris` it doesn't work and remains at its previous value. This is also a problem in current master.

**Comment by phweyland** [comment]
You are right, A bit too fast here ... Let's review this. 
I'm also concerned about potential freeze in case of recursive directory checked.

**Comment by elstoc** [comment]
I was about to say - you should disable recursion whenever selecting a new folder. Combine that with my suggestion and you can also get rid of the parasites.

**Comment by elstoc** [comment]
I still think it's a good idea.

**Comment by phweyland** [comment]
I agree, but need some time to do it properly.

**Comment by phweyland** [comment]
> If I try to change the root to `/home/chris` it doesn't work and remains at its previous value. This is also a problem in current master.

Got it.

**Comment by phweyland** [comment]
some improvements
- on root folder change set it as last directory by default (only if last directory is not this branch)
- add the root folder into the folder pane (and remove the parasitic selection)
- fix bug on folder stack (preventing to select a parent folder)

probably needs more tests ...

**Comment by phweyland** [comment]
hold on... last change messed up ...

**Comment by elstoc** [comment]
You need to un-tick the "recursive directory" tick box when selecting a new root. You're not actually doing a recursion but the box is still ticked. 

When auto-selecting the last chosen subdirectory I'm not sure whether it's better to auto-expand as well or not. On the whole, personally, I prefer to leave it collapsed, but I suppose others might think differently.

We still need to work on how to advertise the process better, because I think the purpose of the root folder is going to confuse a lot of people initially. But this at least means that it will still work for those users who've misunderstood.  Perhaps renaming the "folder" column to "folder navigation" or "folder tree" and adding another verbose tooltip might help?

**Comment by LViatour** [comment]
on linux i removed recursion but could not import a folder. Many of us do not succeed.

**Comment by elstoc** [comment]
> on linux i removed recursion but could not import a folder. Many of us do not succeed.

That was a comment to @phweyland about this PR, but you've still not made your issue clear.

**Comment by LViatour** [comment]
I just want to import in place a folder containing hundreds of raw .nef files.

I go to import in place, then I click at the top right to go to my hard drive and choose the folder. I live my folder, I choose it, I see the raws in the folder, I click on open and the new window is empty without the raws and "import in place" is grayed out.

**Comment by elstoc** [comment]
Then you're misunderstanding how the functionality works. It's understandable - it's not intuitive the first time you use it and we need to better explain that in the UI. Please follow the process I've suggested in the linked issue.

This PR will allow you to use this functionality in the way you're trying to but it's not the optimal way to use it.

**Comment by todd-prior** [comment]
Chris I appreciate your comments but in any way possible I can try this does not work even setting any high level folder as a root so I will wait but for now I have now way to use DT as I cannot access any files. I don't use the database and so each session is an import in place for me.....it worked perfectly fine before I guess I need to get educated but it really looks like the code is broken to me when it offers and empty result to  a folder selection....

**Comment by ptilopteri** [comment]
I to utilize --library :memory: and can import/load via:
  darktable --library :memory /path/to/images

I use full path and can load multiple directories.

linux openSUSE Tumbleweed constantly updating master 
currently a15829eae

**Comment by todd-prior** [comment]
Okay so I'm on my phone but I will try to do a concise work through a bit later. So it does seem to work but if you go into your pictures folder and tunnel down to a folder with no subfolders you see nothing okay that was explained but at least on windows it's
 not releasing that so you can go back and try to set pictures as the root and you can't seem to get it to do that. I just picked C drive as a root and I could tunnel down and import works...so I think the confusion is coming from this weird behaviour for some
 ie if you didn't choose your root properly at the start...





Get Outlook for Android

**Comment by elstoc** [comment]
> not releasing that so you can go back and try to set pictures as the root and you can't seem to get it to do that.

> ie if you didn't choose your root properly at the start...

Yeah that's fixed in the PR and definitely a bug

**Comment by elstoc** [comment]
Can we give the root folder selector button a label? I'd love to call it "library root" or something, which might make it clearer that it's supposed to contain multiple directories. Unfortunately it might make people think it's more than it is - i.e. a restriction that stops them having their images in multiple independent locations.

Perhaps we have "root folder" and "folder tree" as the labels? Hopefully I'll think of something better, but that's the best I can think of for now.

**Comment by todd-prior** [comment]
Thanks for all your feedback..
After playing with it a bit I could see the issue was that I just couldn't work back up the file tree if I went directly down..I think you guys were discussing setting a base directory as is done with Luts ..maybe that is a good idea??  However the current one looks good as long as you can back up the file tree and get the dialogue to repopulate I think it will be good...

**Comment by phweyland** [comment]
> You need to un-tick the "recursive directory" tick box when selecting a new root.

Good catch.

> When auto-selecting the last chosen subdirectory I'm not sure whether it's better to auto-expand as well or not.

If kept collapsed we should not select it and nor show the content either. Simplifies the code as well.
As is, it's consistent. Keeps the tree opened which can be an advantage if it is deep. 


> I to utilize --library :memory: and can import/load via:
> darktable --library :memory /path/to/images

This one doesn't use the import ui, right ?

**Comment by TurboGit** [comment]
I don't think the label "root folder" brings anything ! The icon is already quite common and the tooltip clear.

**Comment by LViatour** [comment]
ok I understood. But I did a lot of tests yesterday so I have a lot of shortcuts to root folders. How to remove them?

**Comment by phweyland** [comment]
> ok I understood. But I did a lot of tests yesterday so I have a lot of shortcuts to root folders. How to remove them?

- clear the darktablerc entry last_root

**Comment by elstoc** [comment]
> I don't think the label "root folder" brings anything ! The icon is already quite common and the tooltip clear.

I don't know. There seems to be a lot of misunderstanding about how this is supposed to be used. People don't read tooltips and we should make the UI as easy as we can without them.

> If kept collapsed we should not select it and nor show the content either. Simplifies the code as well. As is, it's consistent. Keeps the tree opened which can be an advantage if it is deep.

I'm not sure I understand. We should always auto-select the last chosen directory and show its content, except, perhaps, when changing root folder (I think clearing the selection when changing root is better). It's a really common use case for me to repeatedly re-import the same folder (play raw is the main use case).

**Comment by phweyland** [comment]
> > If kept collapsed we should not select it and nor show the content either. Simplifies the code as well. As is, it's consistent. Keeps the tree opened which can be an advantage if it is deep.
> 
> I'm not sure I understand. We should always auto-select the last chosen directory and show its content, except, perhaps, when changing root folder (I think clearing the selection when changing root is better). It's a really common use case for me to repeatedly re-import the same folder (play raw is the main use case).

OK, we are aligned. I may have misinterpreted one of your previous comment.


> I don't know. There seems to be a lot of misunderstanding about how this is supposed to be used.

The main trouble came from the reuse of the last_directory as default last_root reinforced by the bug which prevented to select a higher level in the tree.
I think also that the icon should be enough (if everything works as expected :)).
I remove this.

**Comment by elstoc** [comment]
Ok. Given the bug that's being fixed here, let's get this one merged asap and we can always tweak further later on if we need to.

**Comment by TurboGit** [comment]
Ok, merging !

File: src/libs/import.c [hunk 1]
```diff
@@ -1347,14 +1349,14 @@ static void _import_from_dialog_new(dt_lib_module_t* self)
       folder = g_strdup(folder ? folder : "");
       dt_conf_set_string("ui_last/import_last_root", folder);
       g_free(folder);
+      dt_conf_set_string("ui_last/import_recursive", FALSE);
```
Comment (hlecleme) [inline src/libs/import.c:1352]: Shouldn't this be `dt_conf_set_bool`?

Comment (elstoc) [reply src/libs/import.c:1352]: yes

Comment (phweyland) [reply src/libs/import.c:1352]: Yes !

Comment (phweyland) [reply src/libs/import.c:1352]: sent a fix. Thanks !

**Pull request merged**